File: backend/internal/api/metrics.go
package api

import (
	"net/http"
	"runtime"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/gorilla/websocket"
	"github.com/shirou/gopsutil/v3/cpu"
	"github.com/shirou/gopsutil/v3/disk"
	"github.com/shirou/gopsutil/v3/host"
	"github.com/shirou/gopsutil/v3/mem"
	"github.com/shirou/gopsutil/v3/net"
)

// SystemMetrics represents system resource metrics
type SystemMetrics struct {
	CPU        CPUMetrics     `json:"cpu"`
	Memory     MemoryMetrics  `json:"memory"`
	Disk       DiskMetrics    `json:"disk"`
	Network    NetworkMetrics `json:"network"`
	Uptime     int64          `json:"uptime"`
	Hostname   string         `json:"hostname"`
	OS         string         `json:"os"`
	Platform   string         `json:"platform"`
	LoadAvg    []float64      `json:"loadAvg"`
	Timestamp  int64          `json:"timestamp"`
}

// CPUMetrics represents CPU metrics
type CPUMetrics struct {
	Usage   float64 `json:"usage"`
	Cores   int     `json:"cores"`
	Model   string  `json:"model"`
}

// MemoryMetrics represents memory metrics
type MemoryMetrics struct {
	Total       uint64  `json:"total"`
	Used        uint64  `json:"used"`
	Free        uint64  `json:"free"`
	UsedPercent float64 `json:"usedPercent"`
	Buffers     uint64  `json:"buffers"`
	Cached      uint64  `json:"cached"`
}

// DiskMetrics represents disk metrics
type DiskMetrics struct {
	Total       uint64  `json:"total"`
	Used        uint64  `json:"used"`
	Free        uint64  `json:"free"`
	UsedPercent float64 `json:"usedPercent"`
}

// NetworkMetrics represents network metrics
type NetworkMetrics struct {
	BytesSent   uint64 `json:"bytesSent"`
	BytesRecv   uint64 `json:"bytesRecv"`
	PacketsSent uint64 `json:"packetsSent"`
	PacketsRecv uint64 `json:"packetsRecv"`
}

// HealthCheck returns API health status
func HealthCheck(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{
		"status":    "healthy",
		"timestamp": time.Now().Unix(),
		"version":   "1.0.0",
	})
}

// GetSystemMetrics returns current system metrics
func GetSystemMetrics(c *gin.Context) {
	metrics, err := collectSystemMetrics()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, metrics)
}

// collectSystemMetrics collects all system metrics
func collectSystemMetrics() (*SystemMetrics, error) {
	// CPU
	cpuPercent, _ := cpu.Percent(time.Second, false)
	cpuInfo, _ := cpu.Info()

	cpuUsage := 0.0
	if len(cpuPercent) > 0 {
		cpuUsage = cpuPercent[0]
	}

	cpuModel := "Unknown"
	if len(cpuInfo) > 0 {
		cpuModel = cpuInfo[0].ModelName
	}

	// Memory
	memInfo, err := mem.VirtualMemory()
	if err != nil {
		return nil, err
	}

	// Disk
	diskInfo, err := disk.Usage("/")
	if err != nil {
		return nil, err
	}

	// Network
	netInfo, err := net.IOCounters(false)
	if err != nil {
		return nil, err
	}

	var netMetrics NetworkMetrics
	if len(netInfo) > 0 {
		netMetrics = NetworkMetrics{
			BytesSent:   netInfo[0].BytesSent,
			BytesRecv:   netInfo[0].BytesRecv,
			PacketsSent: netInfo[0].PacketsSent,
			PacketsRecv: netInfo[0].PacketsRecv,
		}
	}

	// Host info
	hostInfo, _ := host.Info()

	return &SystemMetrics{
		CPU: CPUMetrics{
			Usage: cpuUsage,
			Cores: runtime.NumCPU(),
			Model: cpuModel,
		},
		Memory: MemoryMetrics{
			Total:       memInfo.Total,
			Used:        memInfo.Used,
			Free:        memInfo.Free,
			UsedPercent: memInfo.UsedPercent,
			Buffers:     memInfo.Buffers,
			Cached:      memInfo.Cached,
		},
		Disk: DiskMetrics{
			Total:       diskInfo.Total,
			Used:        diskInfo.Used,
			Free:        diskInfo.Free,
			UsedPercent: diskInfo.UsedPercent,
		},
		Network: netMetrics,
		Uptime:     int64(hostInfo.Uptime),
		Hostname:   hostInfo.Hostname,
		OS:         hostInfo.OS,
		Platform:   hostInfo.Platform,
		Timestamp:  time.Now().Unix(),
	}, nil
}

// WebSocket upgrader
var upgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool {
		return true // Allow all origins for development
	},
}

// MetricsWebSocket handles WebSocket connection for real-time metrics
func MetricsWebSocket(c *gin.Context) {
	conn, err := upgrader.Upgrade(c.Writer, c.Request, nil)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	defer conn.Close()

	// Send metrics every second
	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop()

	for range ticker.C {
		metrics, err := collectSystemMetrics()
		if err != nil {
			continue
		}

		if err := conn.WriteJSON(metrics); err != nil {
			break
		}
	}
}

File: src/api/metrics.rs
//! System metrics - CPU, RAM, Disk, Network

use axum::{
    extract::ws::{Message, WebSocket, WebSocketUpgrade},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use serde_json::json;
use sysinfo::{Disks, Networks, System};
use std::time::Duration;

pub async fn health_check() -> impl IntoResponse {
    Json(json!({
        "status": "healthy",
        "timestamp": chrono::Utc::now().timestamp(),
        "version": "2.0.0",
        "engine": "rust/axum"
    }))
}

pub async fn get_system_metrics() -> impl IntoResponse {
    match collect_metrics() {
        Ok(metrics) => (StatusCode::OK, Json(serde_json::to_value(metrics).unwrap())),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({"error": e})),
        ),
    }
}

fn collect_metrics() -> Result<crate::models::SystemMetrics, String> {
    let mut sys = System::new_all();
    sys.refresh_all();
    std::thread::sleep(Duration::from_millis(200));
    sys.refresh_cpu_usage();

    let disks = Disks::new_with_refreshed_list();
    let networks = Networks::new_with_refreshed_list();

    // CPU
    let cpu_usage = sys.global_cpu_info().cpu_usage();
    let cpu_cores = sys.cpus().len();
    let cpu_model = sys
        .cpus()
        .first()
        .map(|c| c.brand().to_string())
        .unwrap_or_else(|| "Unknown".to_string());

    // Memory
    let total_mem = sys.total_memory();
    let used_mem = sys.used_memory();
    let free_mem = sys.free_memory();

    // Disk (root partition)
    let (disk_total, disk_used, disk_free) = disks
        .iter()
        .find(|d| d.mount_point() == std::path::Path::new("/"))
        .map(|d| {
            (
                d.total_space(),
                d.total_space() - d.available_space(),
                d.available_space(),
            )
        })
        .unwrap_or((0, 0, 0));

    // Network
    let (bytes_sent, bytes_recv, pkts_sent, pkts_recv) = networks.iter().fold(
        (0u64, 0u64, 0u64, 0u64),
        |(bs, br, ps, pr), (_name, data)| {
            (
                bs + data.total_transmitted(),
                br + data.total_received(),
                ps + data.total_packets_transmitted(),
                pr + data.total_packets_received(),
            )
        },
    );

    // Host info
    let hostname = System::host_name().unwrap_or_else(|| "unknown".to_string());
    let os_name = System::name().unwrap_or_else(|| "Linux".to_string());
    let os_version = System::os_version().unwrap_or_default();

    // Load average
    let load = System::load_average();

    // Uptime
    let uptime = System::uptime();

    Ok(crate::models::SystemMetrics {
        cpu: crate::models::CpuMetrics {
            usage: cpu_usage,
            cores: cpu_cores,
            model: cpu_model,
        },
        memory: crate::models::MemoryMetrics {
            total: total_mem,
            used: used_mem,
            free: free_mem,
            used_percent: if total_mem > 0 {
                (used_mem as f64 / total_mem as f64) * 100.0
            } else {
                0.0
            },
        },
        disk: crate::models::DiskMetrics {
            total: disk_total,
            used: disk_used,
            free: disk_free,
            used_percent: if disk_total > 0 {
                (disk_used as f64 / disk_total as f64) * 100.0
            } else {
                0.0
            },
        },
        network: crate::models::NetworkMetrics {
            bytes_sent,
            bytes_recv,
            packets_sent: pkts_sent,
            packets_recv: pkts_recv,
        },
        uptime,
        hostname,
        os: os_name,
        platform: os_version,
        load_avg: vec![load.one, load.five, load.fifteen],
        timestamp: chrono::Utc::now().timestamp(),
    })
}

pub async fn metrics_websocket(ws: WebSocketUpgrade) -> impl IntoResponse {
    ws.on_upgrade(handle_metrics_ws)
}

async fn handle_metrics_ws(mut socket: WebSocket) {
    let mut interval = tokio::time::interval(Duration::from_secs(2));

    loop {
        interval.tick().await;

        match collect_metrics() {
            Ok(metrics) => {
                if let Ok(json) = serde_json::to_string(&metrics) {
                    if socket.send(Message::Text(json.into())).await.is_err() {
                        break;
                    }
                }
            }
            Err(_) => continue,
        }
    }
}
